graph/path/bellman_ford_moore.go

```diff
@@ -27,10 +30,63 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
 	path = newShortestFrom(u, nodes)
 	path.dist[path.indexOf[u.ID()]] = 0
 
-	// TODO(kortschak): Consider adding further optimisations
-	// from http://arxiv.org/abs/1111.5414.
-	for i := 1; i < len(nodes); i++ {
-		changed := false
+	// queue to keep track which nodes need to be relaxed
+	// only nodes whose vertex distance changed in the previous iterations need to be relaxed again
+	queue := newBellmanFordQueue()
+	queue.enqueue(u)
+
+	// bool array to keep track whether a node is on the queue or not
+	onQueue := make([]bool, len(nodes))
+	onQueue[path.indexOf[u.ID()]] = true
+
+	n := len(nodes)
+	// the maximum of edges in a graph is |V| * (|V| -1)
+	// which is also the worst case complexity
+	// íf the queue-loop has more iterations than the amount of maximum edges
+	// it indicates that we have a negative cycle
+	maxEdges := n * (n - 1)
+	negativeCycle := false
+
+	loops := 0
+	for queue.len() != 0 {
+		u, err := queue.dequeue()
+		if err != nil {
+			panic(err)
+		}
+		uid := u.ID()
+		onQueue[path.indexOf[uid]] = false
+
+		for _, v := range graph.NodesOf(g.From(uid)) {
+			vid := v.ID()
+			k := path.indexOf[vid]
+			w, ok := weight(uid, vid)
+			if !ok {
+				panic("bellman-ford: unexpected invalid weight")
+			}
+
+			j := path.indexOf[uid]
+			joint := path.dist[j] + w
+			if joint < path.dist[k] {
+				path.set(k, joint, j)
+				index := path.indexOf[vid]
+
+				// check if node is already in the queue
```

Comment:
Punctuation, though as above, this review comment becomes redundant with the queue changes; no comment will be necessary here since `if !queue.has(vid) {` speaks loudly enough.